### ISSUE TITLE
Is Masking Efficient?

### ISSUE THREAD

**Issue by Tonnonssi** [KanghwaSisters/YOLO-Futures#17]
### 📄 설명

PPO 알고리즘에서 액션 마스킹을 하는게 효과적인가? 

### ✅ 작업할 내용

- [x] 알아보자

**Comment by Tonnonssi** [comment]
### 📄 Paper 1  : Implementing action mask in proximal policy optimization (PPO) algorithm (2020)
---
invalid한 행동이란 현재 환경에서 실현 불가능하거나 금지된 행동이다. 예를 들어, 지뢰 찾기에서 이미 까진 타일을 또 누르는 행동은 invalid한 행동이다. 이런 행동을 강화학습에서는 3가지 방식으로 핸들링한다.   

1. invalid한 행동을 무시하고 다음 step으로 넘어간다.  
2. invalid한 행동에 done=True, negative reward를 주어 행동을 피하도록 학습한다. (**Penalty**)
3. invalid한 행동을 마스킹해 강제로 접근하지 못하게 한다. (**Masking**)

본 논문은 PPO에서 마스킹 방법이 패널티 방식보다 빠른 초기 수렴을 가져온다 주장한다. 이론적인 강화학습에서는 2번처럼 음의 보상을 통해 나쁜 선택임을 가르치는 것이 옳다. 하지만 이 방식은 학습 속도를 늦춘다는 단점이 있다. 마스킹 방식은 패널티보다 효율적인 탐험을 유도해 더 빠른 학습을 가능하게 한다.  

<img width="491" height="236" alt="Image" src="https://github.com/user-attachments/assets/e8d0d73d-2501-41ed-812a-426acc5d2057" />

<img width="501" height="230" alt="Image" src="https://github.com/user-attachments/assets/4cc758de-6fea-4306-b86e-3dfaad3f320c" />

### 📄 Paper 2  : A Closer Look at Invalid Action Masking in Policy Gradient Algorithms(2020)
---
본 논문은 정책 그래디언트 알고리즘의 마스킹 적용을 수리적으로 모델링했다. 

$mask$ : 미분가능한 함수로, action space에 대한 logits를 받아 invalid한 값을 죽인다.  
- identity ftn이라서, state에 종속된다. 따라서 $mask(s,x) ̸= mask(s′,x).$ 관계가 성립한다.  

<img width="364" height="109" alt="Image" src="https://github.com/user-attachments/assets/f6bbe067-1f96-4dc3-808f-54a17ab0ee2f" />

- $mask$가 미분가능 함수이기 때문에, 두 수식은 모두 미분가능하다. 따라서 정책 함수 $\pi$는 역전파로 업데이트 가능하다. 


<h3 style="color:red;">✅ PPO에서 액션 마스킹은 효과적이다.</h3>
